File: src/main/java/cn/motui/meican/job/NotificationJob.java
package cn.motui.meican.job;

import cn.motui.meican.model.ui.DateData;
import cn.motui.meican.model.ui.OrderDetail;
import cn.motui.meican.notification.NotificationFactory;
import cn.motui.meican.service.ServiceFactory;
import org.quartz.Job;
import org.quartz.JobExecutionContext;
import org.quartz.JobExecutionException;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;

/**
 * 通知JOB
 *
 * @author it.motui
 * @date 2021-01-24
 */
public class NotificationJob implements Job {

  @Override
  public void execute(JobExecutionContext context) throws JobExecutionException {
    LocalDateTime now = LocalDateTime.now();
    List<DateData> dateDataList = ServiceFactory.dataService().getDateData(now);
    int hour = now.getHour();
    DateData dateData;
    if (hour < 12) {
      dateData = dateDataList.get(0);
    } else {
      dateData = dateDataList.get(1);
    }
    String content = null;
    switch (dateData.getStatus()) {
      case AVAILABLE:
        content = "<p>时间:" + now.format(DateTimeFormatter.ofPattern("yyyy/MM/dd HH:mm"))
            + "</p><p><b>还未订餐</b>，记得订餐哦</p>";
        NotificationFactory.showInfoNotification(content);
        break;
      case ORDER:
        OrderDetail orderDetail = ServiceFactory.dataService().getOrderDetail(dateData.getOrderUniqueId());
        content = "<p>已订餐:</p>" +
            "<p>店铺: " + orderDetail.getRestaurantName() + "</p>" +
            "<p><b>" + orderDetail.getDish().getName() + "</b></p>" +
            "<p>" + orderDetail.getDish().getRemark() + "</p>";
        NotificationFactory.showInfoNotification(content);
        break;
      case CLOSED:
        NotificationFactory.showWarningNotification("[美餐]餐厅关闭,无法订餐");
        break;
      default:
        break;
    }
  }
}

File: src/main/kotlin/cn/motui/meican/job/NotificationJob.kt
package cn.motui.meican.job

import cn.motui.meican.MeiCanBundle
import cn.motui.meican.NOTIFICATIONS_ID
import cn.motui.meican.model.TabStatus
import cn.motui.meican.model.ui.TabData
import cn.motui.meican.util.Notifications
import cn.motui.meican.util.dataService
import org.quartz.Job
import org.quartz.JobExecutionContext
import java.time.LocalDateTime
import java.util.function.Function
import java.util.stream.Collectors

/**
 * 订餐通知JOB
 */
class NotificationJob : Job {
    override fun execute(context: JobExecutionContext?) {
        val title = context?.mergedJobDataMap?.get(jobKey)
        var content: String? = null
        try {
            val tabDataList = dataService.getTabData(LocalDateTime.now())
            val tabData = tabDataList.stream()
                .collect(Collectors.toMap(TabData::title, Function.identity()))[title]
            tabData?.let {
                content = when (it.tabStatus) {
                    TabStatus.AVAILABLE -> MeiCanBundle.message("order.notification").format(it.title)
                    TabStatus.ORDER -> {
                        if (it.corpOrderUser == null) {
                            MeiCanBundle.message("order.notification").format(it.title)
                        } else {
                            val orderDetail = dataService.getOrderDetail(it.corpOrderUser.orderUniqueId)
                            MeiCanBundle.message("order.notification.order.detail").format(
                                orderDetail.restaurantName,
                                orderDetail.dish.name,
                                orderDetail.pickUpMessage
                            )
                        }
                    }
                    else -> MeiCanBundle.message("order.notification.restaurant.close")
                }
            }
        } catch (e: Exception) {
            content = MeiCanBundle.message("order.notification").format("")
        } finally {
            content?.let {
                Notifications.showInfoNotification(
                    NOTIFICATIONS_ID,
                    MeiCanBundle.message("order.notification.title"),
                    it
                )
            }
        }
    }
}
